swagger/smart_transfer_protocol_ws.yaml

```diff
@@ -0,0 +1,421 @@
+            - datasetId
+            - dataFiles
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier     
+            dataFiles:
+              type: array
+              description: List of data file references in this bulk
+              items:
+                $ref: '#/components/schemas/DataFileReference'
+      
+  dataset/pull/data/negotiate/response:
+    publish:
+      operationId: odf.dataset.pull.data.negotiate.response
+      summary: Response on negotiated status of data operations during pull
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullDataNegotiateResponse
+        name: DatasetPullDataNegotiateResponse
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - datasetId
+            - negotiatedDownloads
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier        
+            negotiatedDownloads:
+              type: array
+              description: Negotiated download operations for requested bulk of files
+              items:
+                $ref: '#/components/schemas/NegotiatedDownload'
+
+  dataset/push/request:
+    subscribe:
+      operationId: odf.dataset.push.request
+      summary: Initiates dataset push to the ODF server
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPushRequest
+        name: DatasetPushRequest
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - datasetId
+            - knownHead
+            - newBlocks
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier
+            knownHead:
+              type: string
+              format: multihash
+              description: Hash of the metadata head block known to the pusher
+            newBlocks:
+              $ref: '#/components/schemas/TarOfBlocks'
+
+  dataset/push/request/accepted:
+    publish:
+      operationId: odf.dataset.push.request.accepted
+      summary: Informs about succesfull acceptance of dataset push request
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPushRequestAccepted
+        name: DatasetPushRequestAccepted
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - datasetId
+            - pushRequestId
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier    
+            pushRequestId:
+              type: string
+              format: uuid
+              description: Request identifier of the started push flow
```

Comment:
Same here about the statefulness of the channel. For simplicity I would just assume for now that we can store the state fully on the server.

If we ever need to make this protocol work over a stateless channel - I'd rather do this using an opaque `pushToken` which would let server decide what to store there (e.g. datasetID, expiration time for the push operation etc.)

Reply:
Assuming this would be managed by single co-routine at each side, so the state can be stored on the stack as local variables.